agent/stats/engine_unix.go

```diff
@@ -0,0 +1,104 @@
+//go:build linux
+// +build linux
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//      http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package stats
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"time"
+
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	taskresourcevolume "github.com/aws/amazon-ecs-agent/agent/taskresource/volume"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/csiclient"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/tcs/model/ecstcs"
+
+	"github.com/aws/aws-sdk-go/aws"
+)
+
+func (engine *DockerStatsEngine) getEBSVolumeMetrics(taskArn string) []*ecstcs.VolumeMetric {
+	task, err := engine.resolver.ResolveTaskByARN(taskArn)
+	if err != nil {
+		logger.Error(fmt.Sprintf("Unable to get corresponding task from dd with task arn: %s", taskArn))
+		return nil
+	}
+
+	if !task.IsEBSTaskAttachEnabled() {
+		logger.Debug("Task not EBS-backed, skip gathering EBS volume metrics.", logger.Fields{
+			"taskArn": taskArn,
+		})
+		return nil
+	}
+
+	if engine.csiClient == nil {
```

Comment:
[non-blocking] There isn't any benefit to caching the CSI Client instance in engine struct as each method call on the client instantiates a new connection anyway. Removing the client field from engine will make the engine less stateful.

Reply:
Oh I see, since it's just used to collect the stats, we can just create the a new csi client which will instantiate a new connection regardless and use that to call the CSI driver to get the metrics. Am I understand that correctly?

Reply:
Yeah unless we want to mock the client task engine has for testing.